lib/bookingsync/api/excon_client.rb

```diff
@@ -0,0 +1,359 @@
+    # @param options [Hash] A customizable set of request options.
+    # @return [Array<BookingSync::API::Resource>] Array of resources.
+    def request(method, path, data = nil, options = nil)
+      instrument("request.bookingsync_api", method: method, path: path) do
+        response = call(method, path, data, options)
+        response.respond_to?(:resources) ? response.resources : response
+      end
+    end
+
+    # Make a HTTP GET request to a path with pagination support.
+    #
+    # @param options [Hash]
+    # @option options [Integer] per_page: Number of resources per page
+    # @option options [Integer] page: Number of page to return
+    # @option options [Boolean] auto_paginate: If true, all resources will
+    #   be returned. It makes multiple requestes underneath and joins the results.
+    #
+    # @yieldreturn [Array<BookingSync::API::Resource>] Batch of resources
+    # @return [Array<BookingSync::API::Resource>] Batch of resources
+    def paginate(path, options = {}, &block)
+      instrument("paginate.bookingsync_api", path: path) do
+        auto_paginate = options.delete(:auto_paginate)
+        response = call(:get, path, query: options)
+        data = response.resources.dup
+
+        if (block_given? or auto_paginate) && response.relations[:next]
+          first_request = true
+          loop do
+            if block_given?
+              yield(response.resources)
+            elsif auto_paginate
+              data.concat(response.resources) unless first_request
+              first_request = false
+            end
+            break unless response.relations[:next]
+            response = response.relations[:next].get
+          end
+        end
+
+        data
+      end
+    end
+
+    # Make a HTTP request to given path and returns Response object.
+    #
+    # @param method [Symbol] HTTP verb to use.
+    # @param path [String] The path, relative to {#api_endpoint}.
+    # @param data [Hash] Data to be send in the request's body
+    #   it can include query: key with requests params for GET requests
+    # @param options [Hash] A customizable set of request options.
+    # @return [BookingSync::API::Response] A Response object.
+    def call(method, path, data = nil, options = nil)
+      instrument("call.bookingsync_api", method: method, path: path) do
+        request_params = {
+          :method => method,
+          :headers => { "Authorization" => "Bearer #{token}" }
+        }
+
+        if [:get, :head].include?(method)
+          options = data
+          data = nil
+        end
+
+        if options.has_key?(:query)
+          if options[:query].has_key?(:ids)
+            ids = Array(options[:query].delete(:ids)).join(',')
+            path = "#{path}/#{ids}"
+          end
+          options[:query].keys.each do |key|
+            if options[:query][key].is_a?(Array)
+              options[:query][key] = options[:query][key].join(",")
+            end
+          end
+        end
+
```

Comment:
I think that something similar to old `Client` should be used here - `expand_url(path, options[:uri])`. Otherwise it fails with 404, when we use pagination with block or with `auto_paginate: true` (it returns link to next page as an instance of Addressable::Template, not string)